### ISSUE TITLE
Training time for models

### ISSUE THREAD

**Issue by ThorvaldAagaard** [tensorflow/tensorflow#70247]
### Issue type

Performance

### Have you reproduced the bug with TensorFlow Nightly?

N/A

### Source

binary

### TensorFlow version

2.13

### Custom code

N/A

### OS platform and distribution

x64

### Python version

3.9

### Current behavior?

I have a model created using TF 1.x, that can be trained with a given dataset in about 8 hours.

6 months ago I converted the model to Keras 2, and noted that the training time now suddenly was more than 16 hours, so I ended using the older 1.X model, but used TF 2.13 with support for 1.X.

Now I have upgraded the computer used for training to a i9-14900K with 24 cores (from i9-9900K with 8 cores)

I expected a much faster training, but actually the training is taking about 50% more.

I noticed that my calculations of error during training went down much faster
Old model
![image](https://github.com/tensorflow/tensorflow/assets/10560652/c46e58c8-c1a2-4036-9887-980204d61150)

New model
![image](https://github.com/tensorflow/tensorflow/assets/10560652/f6dfbde0-9eb5-40d5-8b1e-a7dc3951ccb2)

So instead of being concerned with longer training time should I be happy that I can reduce iterations?

Now I am planning to upgrade to Keras 3, and perhaps it will be better.

### Standalone code to reproduce the issue

```shell
If this is expected behavior, where can I find documentation of this?
```

**Comment by sushreebarsa** [comment]
@ThorvaldAagaard Please ensure that your environment and TensorFlow version are fully optimized for your new hardware. Also kindly check if TensorFlow is correctly utilizing all available cores and try with the latest TF version?
For any further queries please post this issue in [Keras](https://github.com/keras-team/keras/issues) repository.

Thank you!

**Comment by sushreebarsa** [comment]
@ThorvaldAagaard Thank you for the update!
TF v1.x is not actively supported and for migration please check [this](https://www.tensorflow.org/guide/migrate) reference. It is highly recommended to use the latest. If you still want to use TF v1.x then kindly post this issue in TF [forum](https://discuss.tensorflow.org/) where there is a larger community to get you the right help. Thank you!

**Comment by ThorvaldAagaard** [comment]
I have tried upgrading to Keras, but notice a decrease in performance.

The real issue is that I only get 50-60 % utilization of my CPU, and I have searched for a solution but without luck.

From my search it seems to be a problem I still will have after switching to Keras.

I can see, that if I train 2 models at the same time, utilization will get to 75%, but still a distance to the expected 100%

**Comment by sushreebarsa** [comment]
@ThorvaldAagaard Could you ensure that data loading and preprocessing are not bottlenecks. You could use multi-threaded data loading and preprocessing with tf.data.Dataset. Please use Prefetch and parallelize the data pipeline as much as possible. Also try experimenting with different batch sizes and let us know?
Thank you!

**Comment by ThorvaldAagaard** [comment]
Remember this is still TF 1.X, so dataloading is done by using a batcher
https://github.com/lorserker/ben/blob/main/src/batcher.py
It could be a bottle neck, but fetching data is way below a millisecond for an iteration.

Increasing batch size seems to lower the total spent time, but even with a very large batch size CPU-usage still is around 50%

I will try different batch sizes, but also LSTM_size and hidden layers have impact on performance.

I have been searching for some best practice, but did not really find anything

My input is 1 mill entries, where each entry is sequence of 8, and the output is then 1 mill * 8 1 hot array.

We started with LSTM 128 and 3 hidden laters
Batch size 64 and 50 epocs (Giving about 3 mill iterations)
Learning rate 0.0005

Changing batch size to 512 is reducing iterations to 1/8 but iteration time is only going up by factor 4, so reducing execution time to the half.

But it seems to lower the quality of the net.

Increasing LSTM to 256 and even 512 seems to improve the net.

So if there is any guidelines (except the trial and error approach) it will be wellcome.

I am trying to upgrade to Keras3, but unfortunately, not that simple, but it looks like the way forward.

**Comment by ThorvaldAagaard** [comment]
Spent the last two days in an attempt to upgrade to TF 2.16, but can't get it to work.

The training seems to run better, but still I can't get up to 100% CPU usage, on the other hand GPU under WSL works fine, so I can  train 3 models at the same time.

Now my problem is: how I can get the solution upgraded to TF2.16. Where to ask for advice?

I have a timeseries with a length of 8, and want a prediction for each of the 8 input.

If I input all 8 sequences the prediction for all 8 sequences is OK.

These predictions are done in a sequence, so I would like to start predicting for the first. In earlier versions I filled the not yet known sequences with zero, and got an ok prediction.

In 2.16 I get a prediction that seems almost unrelated to the first sequence.

Now reading it seems I should mask the the input for sequences unknow, so I have added that layer, and is masking input, but the result is disappointing, and it looks like the NN is not ignoring (Or I don't understand the masking)

I have tested with 1 to 8 input sequences and 8,7 and 6 is OK, 5 and 4 can be used, but the rest is terrible.

I have discussed this a lot with ChatGPT but did not find a solution

Just to shaw an example
Calculated output (one hot array for each sequence)
![image](https://github.com/tensorflow/tensorflow/assets/10560652/aa3a9421-5f50-4aec-8ba4-0ea73f432b7c)

Output If I input all 8 sequences
![image](https://github.com/tensorflow/tensorflow/assets/10560652/a7e18436-f9b1-45b7-88ce-4557f6b9cb5f)

Output if I input first sequence and mask the other 7
![image](https://github.com/tensorflow/tensorflow/assets/10560652/801a6c75-3bc6-4360-bdfa-855c7246b6cd)


I have tried models whit both 8 required sequences and allowed just a single sequence as input, but the result is the same, if giving 1-4 sequences as input result is not useable
